pkg/kubectl/cmd/drain.go

```diff
@@ -227,28 +234,65 @@ func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
 				// gone/missing, not for any other cause.  TODO(mml): something more
 				// sophisticated than this
 				if err == nil && ds != nil {
-					replicated = true
+					// Otherwise, treat daemonset-managed pods as unmanaged since
+					// DaemonSet Controller currently ignores the unschedulable bit.
+					// FIXME(mml): Add link to the issue concerning a proper way to drain
+					// daemonset pods, probably using taints.
+					daemonset_pod = true
 				}
 			}
 		}
-		if replicated || o.Force {
-			pods = append(pods, pod)
-		}
-		if !replicated {
+		switch {
+		case daemonset_pod:
+			daemonSetPodNames = append(daemonSetPodNames, pod.Name)
+		case !replicated:
 			unreplicatedPodNames = append(unreplicatedPodNames, pod.Name)
+			if o.Force {
+				pods = append(pods, pod)
+			}
+		default:
+			pods = append(pods, pod)
 		}
 	}
 
-	if len(unreplicatedPodNames) > 0 {
-		joined := strings.Join(unreplicatedPodNames, ", ")
-		if !o.Force {
-			return pods, fmt.Errorf("refusing to continue due to pods managed by neither a ReplicationController nor a DaemonSet: %s (use --force to override)", joined)
+	if (!o.IgnoreDaemonsets && len(daemonSetPodNames) > 0) || (!o.Force && len(unreplicatedPodNames) > 0) {
```

Comment:
Honestly, this logic has just become too complicated for a reader to easily understand. Can you just write a single 16-arm if/else if clause that covers all the possible values of len(daemonSetPodNames), len(unreplicatedPodNames), o.IgnoreDaemonsets, and o.Force?

Also, in the cases where you return error as second argument, can you return empty pod list to make it clearer that it means no pods will be deleted?